roles/ceph-ansible/tasks/main.yml

```diff
@@ -100,16 +100,46 @@
     regexp: "^#journal_size: 0.*"
     line: 'journal_size: 5120'
 
-- name: calculate network prefix
-  shell: "ipcalc -p 0.0.0.0 {{ hostvars[groups['mons'][0]].ansible_default_ipv4['netmask'] }} | sed 's|PREFIX=||'"
-  register: net_prefix
-  changed_when: false
+- set_fact:
+    pub_network: "{{ hostvars[groups['mons'][0]].ansible_default_ipv4.network }}"
+    pub_prefix: "{{ ['0.0.0.0/', hostvars[groups['mons'][0]].ansible_default_ipv4.netmask] | join('') | ipaddr('prefix') }}"
 
-- name: Set public_network to ansible_default_ipv4['address']/16
+- name: Set public_network to ansible_default_ipv4['network']/prefix
   lineinfile:
     dest: "{{ base_dir }}/ceph-ansible/group_vars/all.yml"
     regexp: "^#public_network: 0.0.0.0/0.*"
-    line: "public_network: {{ hostvars[groups['mons'][0]].ansible_default_ipv4['network'] }}/{{ net_prefix.stdout }}"
+    line: "public_network: {{ pub_network }}/{{ pub_prefix }}"
+
+- name: Filter additional NICs (exclude ansible default interface and lo)
+  set_fact:
+    additional_network_interfaces: "{{ hostvars[groups['osds'][0]].ansible_interfaces | difference(['lo', hostvars[groups['osds'][0]].ansible_default_ipv4.interface]) }}"
+- name: Prepare empty list for additional_networks
+  set_fact:
+    additional_networks: []
+- name: Filter configured additional interfaces
+  set_fact:
+    additional_networks: "{{ additional_networks }} + [ {{ hostvars[groups['osds'][0]]['ansible_'+item].ipv4 }} ]"
+  when:
+   - "hostvars[groups['osds'][0]]['ansible_'+item] is defined"
+   - "hostvars[groups['osds'][0]]['ansible_'+item].ipv4 is defined"
+  with_items: "{{ additional_network_interfaces }}"
+- debug: var=additional_networks
+- name: Use first additional interface as data_network/data_prefix
+  set_fact:
+    data_network: "{{ additional_networks.0.network }}"
+    data_prefix: "{{ ['0.0.0.0/', additional_networks.0.netmask] | join('') | ipaddr('prefix') }}"
+  when: additional_networks.0 is defined
```

Comment:
The code in this playbook is quite general - I've tested the particular code on few various scenarios with 1, 2, 3 and 5 NICs (or something like that), with various combinations of configured and not configured NICs and it seems to work quite smoothly and predictably. Only one exception is, when more than two NICs are configured, it is not possible to automatically distinguish which network should be the `data_network`, so the first one from the remainings is used.

About our current infrastructure, we are ready for testing with one and two networks (See the `SECOND_NETWORK` parameter when building new cluster). I've checked both installations and it works well.